exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillFileSystem.java

```diff
@@ -65,46 +62,105 @@
 import com.google.common.collect.Maps;
 
 /**
- * DrillFileSystem is the wrapper around the actual FileSystem implementation.
+ * DrillFileSystem is the wrapper around the actual FileSystem implementation. The {@link DrillFileSystem} is
+ * immutable.
  *
  * If {@link org.apache.drill.exec.ops.OperatorStats} are provided it returns an instrumented FSDataInputStream to
  * measure IO wait time and tracking file open/close operations.
  */
 public class DrillFileSystem extends FileSystem implements OpenFileTracker {
   static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DrillFileSystem.class);
   private final static boolean TRACKING_ENABLED = AssertionUtil.isAssertionsEnabled();
+  private final static DrillFileSystemCache CACHE = new DrillFileSystemCache();
 
+  public static final String FS_DEFAULT_NAME = "fs.default.name";
   public static final String UNDERSCORE_PREFIX = "_";
   public static final String DOT_PREFIX = ".";
 
   private final ConcurrentMap<DrillFSDataInputStream, DebugStackTrace> openedFiles = Maps.newConcurrentMap();
 
+  private final Configuration fsConf;
   private final FileSystem underlyingFs;
   private final OperatorStats operatorStats;
   private final CompressionCodecFactory codecFactory;
 
+  private boolean initialized = false;
+
   public DrillFileSystem(Configuration fsConf) throws IOException {
     this(fsConf, null);
   }
 
   public DrillFileSystem(Configuration fsConf, OperatorStats operatorStats) throws IOException {
-    this.underlyingFs = FileSystem.get(fsConf);
+    // Configuration objects are mutable, and the underlying FileSystem object may directly use a passed in Configuration.
+    // In order to avoid scenarios where a Configuration can change after a DrillFileSystem is created, we make a copy
+    // of the Configuration.
+    this.fsConf = new Configuration(fsConf);
+    normalize(fsConf);
```

Comment:
Why not use `Configuration.setDeprecateProperties` instead ? See [here](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configuration.java#L577)

Reply:
Thanks for showing me this, never knew about it. 

I think setDeprecateProperties does some of what we need, but would still have a problem. Consider the following input and output mappings for setDeprecateProperties.

(fs.default.name: val1 ) -> (fs.default.name: val1 )
(fs.default.name: val1, fs.defaultFS: val2 ) -> (fs.default.name: val1, fs.defaultFS: val2 )
(fs.default.name: val1, fs.defaultFS: val1 ) -> (fs.default.name: val1, fs.defaultFS: val1 )

Ideally we would want all this outputs to be the same and we would want a warning if our given config has different values for fs.default.name and fs.defaultFS.

The normalize method handles this cases. I am trying to think of a way to use setDeprecateProperties to replace some of the existing code in normalize. But I can't think of a way to use setDeprecateProperties with the extra case handling to make the code any simpler than what it is now. 

If you have any ideas on a way we could use setDeprecateProperties and handle the other cases described above more simply than the current normalize method, let me know and I will do it. Otherwise if we can't think of anything now, I'd prefer to keep the normalize method as is.

Reply:
Please see comment below `fs.defaultFS` is new property and f`s.default.name` is the deprecated one. I can look more on how to actually use deprecatedContext but my assumption was internally FileSystem library must be initializing the mapping between deprecated and new properties and use new properties value correctly. Also `fs.defaultFS` will always be set since it has a default value of `file:///`

Reply:
Wow, I had it completely mixed up. You are right, thanks for catching this before it went in! I will fix the normalize method, and revisit trying to use setDeprecateProperties.